File: src/main/java/com/my/admin/app/commerce/model/repository/ProductRepository.java
package com.my.admin.app.commerce.model.repository;

import com.black.food.manager.model.entity.Product;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.Set;

@Repository
public interface ProductRepository extends JpaRepository<Product, Long> {
    Set<Product> findByProductTypeId(Long productTypeId);

    @Query("SELECT topping\n"
            + "FROM  ProductTopping pt\n"
            + "JOIN  Product mainDish on mainDish.id = pt.mainDishId\n"
            + "JOIN  Product topping on topping.id = pt.toppingId\n"
            + "WHERE mainDish.id = :mainDishId")
    Set<Product> findProductToppingsByMainDishId(@Param("mainDishId") Long mainDishId);

    @Query("SELECT product\n"
            + "FROM  Product product\n"
            + "WHERE (:id is null or product.id = :id) "
            + "and (:productTypeId is null or product.productType.id = :productTypeId) "
            + "and (:nameCode is null or (product.name like %:nameCode% or product.code like %:nameCode%)) ")
    Set<Product> productFilter(@Param("id") Long id, @Param("productTypeId") Long productTypeId, @Param("nameCode") String nameCode);
}

File: src/main/kotlin/com/bk/reactive/app/commerce/admin/my/repository/ProductRepository.kt
package com.bk.reactive.app.commerce.admin.my.repository

import com.bk.reactive.app.commerce.admin.my.model.document.Product
import org.springframework.data.mongodb.repository.Query
import org.springframework.data.mongodb.repository.ReactiveMongoRepository
import org.springframework.stereotype.Repository
import reactor.core.publisher.Flux

@Repository
interface ProductRepository : ReactiveMongoRepository<Product, String> {
    @Query("{'productType._id' : :#{#productTypeId}}")
    fun findByProductTypeId(productTypeId: String): Flux<Product>

    //use SpEL syntax
    @Query(
        "{ \$and : ["
                + " {\$or : [ { \$where: \"':#{#productId}' == 'null'\" } , {_id: :#{#productId}} ]},"
                + " {\$or : [ { \$where: \"':#{#productTypeId}' == 'null'\" } , {'productType._id': :#{#productTypeId}} ]}"
                + " ]}"
    ) //+ "and (:nameCode is null or (product.name like %:nameCode% or product.code like %:nameCode%)) ")
    fun productFilter(productId: String?, productTypeId: String?): Flux<Product>
}